File: backend/src/main/java/cn/etarch/mao/weixin/service/ContextTokenRepository.java
package cn.etarch.mao.weixin.service;

import cn.etarch.mao.weixin.entity.WeixinChannelContextToken;
import cn.etarch.mao.weixin.mapper.WeixinChannelContextTokenMapper;
import com.baomidou.mybatisplus.core.conditions.query.LambdaQueryWrapper;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

@Slf4j
@Service
@RequiredArgsConstructor
public class ContextTokenRepository {

    private final WeixinChannelContextTokenMapper contextTokenMapper;

    /**
     * 保存或更新context_token
     */
    public void saveOrUpdate(String accountId, String wxUserId, String token) {
        LambdaQueryWrapper<WeixinChannelContextToken> wrapper = new LambdaQueryWrapper<>();
        wrapper.eq(WeixinChannelContextToken::getAccountId, accountId)
                .eq(WeixinChannelContextToken::getWxUserId, wxUserId);

        WeixinChannelContextToken existing = contextTokenMapper.selectOne(wrapper);
        if (existing != null) {
            existing.setToken(token);
            contextTokenMapper.updateById(existing);
        } else {
            WeixinChannelContextToken newToken = new WeixinChannelContextToken();
            newToken.setAccountId(accountId);
            newToken.setWxUserId(wxUserId);
            newToken.setToken(token);
            contextTokenMapper.insert(newToken);
        }

        log.debug("保存context_token成功, accountId={}, wxUserId={}", accountId, wxUserId);
    }

    /**
     * 获取最新的context_token
     */
    public String getLatestToken(String accountId, String wxUserId) {
        LambdaQueryWrapper<WeixinChannelContextToken> wrapper = new LambdaQueryWrapper<>();
        wrapper.eq(WeixinChannelContextToken::getAccountId, accountId)
                .eq(WeixinChannelContextToken::getWxUserId, wxUserId)
                .last("LIMIT 1");

        WeixinChannelContextToken token = contextTokenMapper.selectOne(wrapper);
        return token != null ? token.getToken() : null;
    }

    /**
     * 删除指定账号的所有context_token
     */
    public void deleteByAccountId(String accountId) {
        LambdaQueryWrapper<WeixinChannelContextToken> wrapper = new LambdaQueryWrapper<>();
        wrapper.eq(WeixinChannelContextToken::getAccountId, accountId);
        contextTokenMapper.delete(wrapper);
        log.info("删除账号的所有context_token, accountId={}", accountId);
    }
}

File: backend-ts/src/weixin/context-token.repository.ts
import type { Db } from '../db/db.js';
import type { WeixinChannelContextToken } from './types.js';

const NOT_DELETED = 'deleted = 0';

export class ContextTokenRepository {
  constructor(private readonly db: Db) {}

  async saveOrUpdate(accountId: string, wxUserId: string, token: string): Promise<void> {
    const existing = await this.db.queryOne<WeixinChannelContextToken>(
      `SELECT * FROM weixin_channel_context_token WHERE account_id = ? AND wx_user_id = ? AND ${NOT_DELETED} LIMIT 1`,
      [accountId, wxUserId],
    );
    if (existing != null && existing.id != null) {
      existing.token = token;
      await this.db.updateById('weixin_channel_context_token', existing.id, { token });
    } else {
      await this.db.insert('weixin_channel_context_token', {
        accountId,
        wxUserId,
        token,
        deleted: 0,
      });
    }
  }

  async getLatestToken(accountId: string, wxUserId: string): Promise<string | null> {
    const row = await this.db.queryOne<WeixinChannelContextToken>(
      `SELECT * FROM weixin_channel_context_token WHERE account_id = ? AND wx_user_id = ? AND ${NOT_DELETED} LIMIT 1`,
      [accountId, wxUserId],
    );
    return row?.token ?? null;
  }

  async deleteByAccountId(accountId: string): Promise<void> {
    await this.db.execute(
      `UPDATE weixin_channel_context_token SET deleted = 1 WHERE account_id = ? AND ${NOT_DELETED}`,
      [accountId],
    );
    console.info(`删除账号的所有context_token, accountId=${accountId}`);
  }

  findByAccountId(accountId: string): Promise<WeixinChannelContextToken[]> {
    return this.db.query<WeixinChannelContextToken>(
      `SELECT * FROM weixin_channel_context_token WHERE account_id = ? AND ${NOT_DELETED} ORDER BY updated_at DESC, id DESC`,
      [accountId],
    );
  }
}
